File: backend/gallicagram.py
from io import StringIO
from dataclasses import dataclass

from fastapi import HTTPException
from models import PeriodRecord, Ticket
import aiohttp
from date import Date
from typing import Callable, Literal
from urllib.error import HTTPError
import pandas as pd


async def get(args: Ticket, on_no_records_found: Callable):
    """Get records from Pyllica. Interpret 500 error as no records found."""
    if args.year is None:
        debut = 1789
    else:
        debut = int(Date(args.year).year)
    if args.end_year is None:
        fin = 1950
    else:
        fin = int(Date(args.end_year).year)
    try:
        if args.link_term:
            periods = await get_contain_data(
                mot1=args.term,
                mot2=args.link_term,
                corpus=args.source,
                resolution="mois",
                debut=debut,
                fin=fin,
            )
        else:
            periods = await get_gram_data(
                gram=args.term,
                corpus=args.source,
                resolution="mois",
                debut=debut,
                fin=fin,
            )
    except HTTPError:
        on_no_records_found()
        return
    if all(periods.ratio == 0):
        on_no_records_found()
        return
    return convert_data_frame_to_grouped_record(periods)


def convert_data_frame_to_grouped_record(frame):
    if "mois" not in frame.columns:
        dates = frame.annee
    else:
        dates = frame.apply(lambda row: f"{row.annee}-{row.mois:02}", axis=1)
    return (
        PeriodRecord(
            _date=Date(date),
            count=count,
            term=term,
        )
        for date, count, term in zip(dates, frame.ratio, frame.gram)
    )


async def get_contain_data(
    mot1: str,
    mot2: str,
    corpus: Literal["lemonde", "livres", "presse"] = "presse",
    debut=1789,
    fin=1950,
    resolution: Literal["default", "annee", "mois"] = "default",
):
    async with aiohttp.ClientSession() as session:
        async with session.get(
            "https://shiny.ens-paris-saclay.fr/guni/contain",
            params={
                "corpus": corpus,
                "mot1": mot1.lower(),
                "mot2": mot2.lower(),
                "from": debut,
                "to": fin,
            },
        ) as response:
            if response.status != 200:
                raise HTTPException(
                    status_code=503, detail="Could not connect to Gallicagram! Egads!"
                )
            return await parse_df(
                response=response, resolution=resolution, corpus=corpus
            )


async def get_gram_data(
    gram: str,
    corpus: Literal["lemonde", "livres", "presse"] = "presse",
    debut=1789,
    fin=1950,
    resolution: Literal["default", "annee", "mois"] = "default",
):
    async with aiohttp.ClientSession() as session:
        async with session.get(
            "https://shiny.ens-paris-saclay.fr/guni/query",
            params={
                "corpus": corpus,
                "mot": gram.lower(),
                "from": debut,
                "to": fin,
            },
        ) as response:
            if response.status != 200:
                raise HTTPException(
                    status_code=503, detail="Could not connect to Gallicagram! Egads!"
                )
            return await parse_df(
                response=response, resolution=resolution, corpus=corpus
            )


async def parse_df(
    response: aiohttp.ClientResponse,
    resolution: Literal["default", "annee", "mois"] = "default",
    corpus: Literal["lemonde", "livres", "presse"] = "presse",
):
    df = pd.read_csv(StringIO(await response.text()))
    if resolution == "mois" and corpus != "livres":
        df = (
            df.groupby(["annee", "mois", "gram"])
            .agg({"n": "sum", "total": "sum"})
            .reset_index()
        )
    if resolution == "annee":
        df = (
            df.groupby(["annee", "gram"])
            .agg({"n": "sum", "total": "sum"})
            .reset_index()
        )

    def calc_ratio(row):
        if row.total == 0:
            return 0
        return row.n / row.total

    df["ratio"] = df.apply(lambda row: calc_ratio(row), axis=1)
    return df

File: frontend/src/app/gallicagram.ts
import { DataFrame, toJSON } from "danfojs-node";
import * as Papa from "papaparse";
import { GraphData } from "./components/models/dbStructs";

type CorpusType = "lemonde" | "livres" | "presse";
type ResolutionType = "default" | "annee" | "mois";

type Ticket = {
  term: string;
  year?: number;
  end_year?: number;
  codes?: string[];
  source: CorpusType;
  grouping?: ResolutionType;
  smoothing?: number;
};

type GallicaGramRow = {
  n_sum: number;
  gram: string;
  annee: string;
  mois: string;
  total_sum: number;
  ratio: number;
};

export async function getSeries(
  { term, year, end_year, grouping, smoothing, source }: Ticket,
  onNoRecordsFound: () => void
): Promise<GraphData | undefined> {
  if (!term) {
    return;
  }
  const debut = year ?? 1789;
  const fin = end_year ?? 1950;

  try {
    const response = await fetch(
      `https://shiny.ens-paris-saclay.fr/guni/query?corpus=${source}&mot=${term}&from=${debut}&to=${fin}&resolution=${grouping}`
    );
    const stringResponse = await response.text();
    const parsedCSV = Papa.parse(stringResponse, { header: true });
    let dataFrame = new DataFrame(parsedCSV.data, {
      columns: ["n", "gram", "annee", "mois", "total"],
    });
    let groupedFrame: DataFrame | null = null;
    if (grouping === "mois" && source !== "livres") {
      groupedFrame = dataFrame
        .groupby(["annee", "mois", "gram"])
        .agg({ n: "sum", total: "sum" })
        .resetIndex();
    }
    if (grouping === "annee") {
      groupedFrame = dataFrame
        .groupby(["annee", "gram"])
        .agg({ n: "sum", total: "sum" })
        .resetIndex();
    }
    let rows = toJSON(groupedFrame, { format: "column" }) as GallicaGramRow[];
    rows = rows.map((row) => ({
      ...row,
      ratio: calculateRatio(row),
    }));
    return {
      data: rows.map((row) => [
        new Date(parseInt(row.annee), parseInt(row.mois) - 1).getTime(),
        row.ratio,
      ]),
      name: term,
    };
  } catch (error) {
    onNoRecordsFound();
    throw error;
  }
}

function calculateRatio(row: GallicaGramRow) {
  if (!row.total_sum) {
    return 0;
  }
  return row.n_sum / row.total_sum;
}
